cmd/crypto/kms.go

```diff
@@ -67,6 +67,48 @@ func (c Context) WriteTo(w io.Writer) (n int64, err error) {
 	return n + int64(nn), err
 }
 
+// AppendTo appends the context in a canonical from to dst.
+//
+// AppendTo sorts the context keys and writes the sorted
+// key-value pairs as canonical JSON object to w.
+func (c Context) AppendTo(dst []byte) (output []byte) {
```

Comment:
I think we can simplify the API to `func (c Context) MarshalText() ([]byte, error)` and implement [`encoding.TextMarshaler`](https://golang.org/pkg/encoding/#TextMarshaler) - (it never returns an error).

I don't think we need to append to an existing slice?!

Reply:
It was mostly to keep it flexible. When it receives a destination we are not forcing an allocation, but leaving that to the caller.

If we should need to do 1000 we can reuse the destination. Similar if we should need to serialize it we can do it without copying the result.

I know it isn't being used, but it is a "might as well" in this case.

Reply:
https://blog.filippo.io/efficient-go-apis-with-the-inliner/

Reply:
@aead Yes. Do you have a point with that link?

Reply:
Yes, by implementing `encoding.Marshaler` like this:
```go
func (c Context) MarshalText() ([]byte, error) {
      text := make(...)
      return c.marshalText(text), nil
}
```
you get both - a "consumer-friendly" API (+ being a `encoding.Marshaler`) + no heap allocation due to the inliner.

Reply:
@aead You don't get inlining when using interfaces and values will always be seen to escape when used as an interface. The "mid-stack inliner" is unreliable and relies on an arbitrary function score.

Real world:
```
func BenchmarkContext_AppendTo(b *testing.B) {
	c := Context{"0": "1", "-": "2", ".": "#"}
	dst := make([]byte, 0, b.N*1024)
	b.ReportAllocs()
	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		dst = c.AppendTo(dst)
	}
}

func BenchmarkContext_MarshalText(b *testing.B) {
	c := Context{"0": "1", "-": "2", ".": "#"}
	dst := make([]byte, 0, b.N*1024)
	b.ReportAllocs()
	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		x, _ := c.MarshalText()
		dst = append(dst, x...)
	}
}
```

The marshal func:
```
func (c Context) MarshalText() (output []byte, err error) {
	dst := make([]byte, 0, 1024)
	if len(c) == 0 {
		return append(dst, '{', '}'), nil
	}

	// No need to copy+sort
	if len(c) == 1 {
		dst = append(dst, '{')
		for k, v := range c {
			dst = append(dst, '"')
			dst = append(dst, []byte(k)...)
			dst = append(dst, '"', ':', '"')
			dst = append(dst, []byte(v)...)
			dst = append(dst, '"')
		}
		return append(dst, '}'), nil
	}

	sortedKeys := make([]string, 0, len(c))
	for k := range c {
		sortedKeys = append(sortedKeys, k)
	}
	sort.Strings(sortedKeys)

	dst = append(dst, '{')
	for i, k := range sortedKeys {
		dst = append(dst, '"')
		dst = append(dst, []byte(k)...)
		dst = append(dst, '"', ':', '"')
		dst = append(dst, []byte(c[k])...)
		dst = append(dst, '"')
		if i < len(sortedKeys)-1 {
			dst = append(dst, ',')
		}
	}
	return append(dst, '}'), nil
}
```

Results: 
```
BenchmarkContext_AppendTo-32    	 3726744	       321 ns/op	      80 B/op	       2 allocs/op
BenchmarkContext_MarshalText-32    	 1892852	       634 ns/op	    1104 B/op	       3 allocs/op
```

The 80B/op comes from the sorting.

Reply:
I could add

```
func (c Context) MarshalText() (output []byte, err error) {
	return c.AppendTo(make([]byte, 0, 128)), nil
}
```
Though I don't really see a point.